### DISCUSSION TITLE
How to create a new DataType with custom visualization?

### DISCUSSION THREAD

**Discussion by hstiffel** [NationalSecurityAgency/ghidra#5660] [Q&A]
Hi,

I want to write a new data type with its own visualization in the listing window, similar to the existing builtin GIF and PNG datatypes. Seeing the source code of those two datatypes, it doesn't look too difficult. But how do you deliver the new datatype to the user? Is there a user friendly way to install new datatypes (e.g. like plugins)? Datatypes don't seem to be covered by the GhidraDev module wizard for Eclipse.

In the documentation of ghidra.program.model.data.BuiltIn, I have found this information:

> NOTE: ALL DATATYPE CLASSES MUST END IN "DataType". If not, the ClassSearcher will not find them. Base class for built-in Datatypes. A built-in data type is searched for in the classpath and added automatically to the available data types in the data type manager.

What about non-builtin datatypes?

**Comment by dev747368** [comment]
re: discovery of your data type class.  As long as your data type's class name ends in the right suffix (eg. "DataType"), it will be discovered at runtime.... if your module / extension has a special manifest file ```<extension_root>/data/ExtensionPoint.manifest``` that lists that suffix.
As an example, take a look at that file in the "Framework SoftwareModeling" project and you will see that it has the "DataType" suffix listed.

**Reply by hstiffel** [reply]
Thanks a lot. The term "ExtensionPoint" in your answer helped me to find more information in the Internet about the topic. But it's quite scattered (a video by Mike Bell, an expensive online course, a blog about the brainfuck CPU for Ghidra,...). Is there an "official" place that explains the extension mechanism of Ghidra? (sorry if I have missed it)

**Comment by dragonmacher** [comment]
There is no single place that documents this, IIRC.   You can look at the javadoc for the relevant classes to get started.   The `Generic` module's `ghidra.util.classfinder` package has the classes that handle this part of the system.   

```
/**
 * NOTE: ExtensionPoint logistics have changed! It is no longer sufficient to
 * implement ExtensionPoint in order for the ClassSearcher to dynamically pick
 * up your class. Your class also needs to conform to a class name suffix rule.
 * The modules included in your application can have a file named
 * "{ModuleRoot}/data/ExtensionPoint.manifest". This file contains (one per
 * line) the suffixes that should be checked for inclusion into the class
 * searching. IF YOUR EXTENSION POINT DOES NOT HAVE A SUFFIX INDICATED IN ONE OF
 * THESE FILES, IT WILL NOT BE AUTOMATICALLY DISCOVERED.
 * 
 * This is a marker interface used to mark classes and interfaces that Ghidra
 * will automatically search for and load.
 */
public interface ExtensionPoint {
	// Marker interface
}
```

and

```
/**
 * This class is a collection of static methods used to discover classes that implement a
 * particular interface or extend a particular base class.
 * <p>
 * <b>Warning: Using the search feature of this class will trigger other classes to be loaded.
 * Thus, clients should not make calls to this class inside of static initializer blocks</b>
 *
 * <p>Note: if your application is not using a module structure for its release build, then
 * your application must create the following file, with the required entries,
 * in order to find extension points:
 * <pre>
 * 	&lt;install dir&gt;/data/ExtensionPoint.manifest
 * </pre>
 *
 */
public class ClassSearcher {
```

In your case, you do not need to add the `DataType` suffix to any `ExtensionPoint.manifest` file, since the system already does that.    

As far as how the extension point system works, there really isn't any other special  knowledge.  It is a very simple system that uses classpath scanning to find all classes and then filters those results by the extension point name suffixes defined in the `ExtensionPoint.manifest` files.
